### ISSUE TITLE
Create something like RuboCop's node matcher macros

### ISSUE THREAD

**Issue by mvz** [troessner/reek#1288]
As an experiment, I implemented part of NilCheck using the RuboCop system. I found the node patterns and matcher macros very nice and useful.

More information here: https://github.com/bbatsov/rubocop/blob/f5fed30d4106c8aff7dd27ffef947def530deae1/manual/node_pattern.md

**Comment by Drenmi** [comment]
They are tremendously useful. 🙂 

There are some pending features for the node matchers, for example optional fragments, that currently no-one is working on. If @bbatsov is up for it, perhaps it can be extracted into a separate gem, and contributions can be made from more directions. WDYT?

**Comment by pocke** [comment]
Yes, I would like to extract it into a separate gem. And I tried to extract it, but it has been stopped because I was very busy. https://github.com/pocke/node_pattern
It rewrites node pattern with kpeg for readability. https://github.com/pocke/node_pattern/pull/1

**Comment by bbatsov** [comment]
I'm not opposed to the idea. A lot of things can be extracted from RuboCop, but I simply never got to doing so. Much of the AST node extensions are also something that'd be useful to anyone using underlying `ast` gem. Probably at some point we can move RuboCop to some organization and place there some related projects. 

@pocke Cool!

**Comment by mvz** [comment]
Having this extracted to a separate gem would be great for us. @pocke do you think your attempt can be continued or is it too outdated?

**Comment by bbatsov** [comment]
I see it's just 21 days old. I would hardly call this outdated. 😄

**Comment by mvz** [comment]
Oh wow, I wouldn't have guessed. My timeframe for concluding 'stopped because too busy' is a bit longer than @pocke's, it seems 😆.
